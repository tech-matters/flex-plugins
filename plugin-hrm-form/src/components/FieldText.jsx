import React from 'react';
import PropTypes from 'prop-types';

import { StyledInput, StyledLabel, ErrorText, TextField } from '../Styles/HrmStyles';
import RequiredAsterisk from './RequiredAsterisk';
import { fieldType } from '../types';

const FieldText = ({ id, label, placeholder, field, rows, handleBlur, handleChange, handleFocus, ...rest }) => (
  <TextField {...rest}>
    <StyledLabel htmlFor={id}>
      {label}
      <RequiredAsterisk field={field} />
    </StyledLabel>
    <StyledInput
      id={id}
      placeholder={placeholder}
      error={field.error !== null}
      value={field.value}
      multiline={Boolean(rows)}
      rows={rows}
      onBlur={handleBlur}
      onChange={handleChange}
      onFocus={handleFocus}
    />
    {field.error && <ErrorText>{field.error}</ErrorText>}
  </TextField>
);

FieldText.displayName = 'FieldText';
FieldText.defaultProps = {
  rows: undefined,
};

FieldText.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  field: fieldType.isRequired,
  rows: PropTypes.number,
  handleBlur: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  handleFocus: PropTypes.func.isRequired,
};
FieldText.defaultProps = {
  label: '',
  placeholder: '',
};

export default FieldText;
