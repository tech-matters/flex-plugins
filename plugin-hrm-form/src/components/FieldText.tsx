/* eslint-disable react/prop-types */
import React from 'react';

import { StyledInput, StyledLabel, ErrorText, TextField } from '../styles/HrmStyles';
import RequiredAsterisk from './RequiredAsterisk';
import { FormFieldType } from './common/forms/types';

type OwnProps = {
  id: string;
  label?: string | JSX.Element;
  placeholder?: string;
  field: FormFieldType;
  rows?: number;
  // Someday we should make this consistent with DefaultEventHandlers in /components/common/forms/types.ts
  handleBlur: React.FocusEventHandler<HTMLDivElement | HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>;
  handleChange: React.ChangeEventHandler<HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>;
  handleFocus: React.FocusEventHandler<HTMLDivElement | HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>;
};

const FieldText: React.FC<OwnProps> = ({
  id,
  label,
  placeholder,
  field,
  rows,
  handleBlur,
  handleChange,
  handleFocus,
  ...rest
}) => (
  <TextField {...rest}>
    <StyledLabel htmlFor={id}>
      {label}
      <RequiredAsterisk field={field} />
    </StyledLabel>
    <StyledInput
      id={id}
      placeholder={placeholder}
      error={field.error !== null}
      value={field.value}
      multiline={Boolean(rows)}
      rows={rows}
      onBlur={handleBlur}
      onChange={handleChange}
      onFocus={handleFocus}
    />
    {field.error && <ErrorText>{field.error}</ErrorText>}
  </TextField>
);

FieldText.displayName = 'FieldText';

FieldText.defaultProps = {
  label: '',
  placeholder: '',
  rows: undefined,
};

export default FieldText;
