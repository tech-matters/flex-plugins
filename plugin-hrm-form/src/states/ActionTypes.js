export const HANDLE_BLUR = 'HANDLE_BLUR';
export const HANDLE_FOCUS = 'HANDLE_FOCUS';
export const HANDLE_CHANGE = 'HANDLE_CHANGE';
export const INITIALIZE_CONTACT_STATE = 'INITIALIZE_CONTACT_STATE';
export const REMOVE_CONTACT_STATE = 'REMOVE_CONTACT_STATE';
export const SAVE_END_MILLIS = 'SAVE_END_MILLIS';
export const SAVE_CONTACT_STATE = 'SAVE_CONTACT_STATE';
