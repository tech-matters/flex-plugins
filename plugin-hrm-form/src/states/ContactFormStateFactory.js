export const ValidationType = {
  REQUIRED: 'REQUIRED', // Will not be applied if in the callerInformation tab and callType is not caller.  Will not be applied when callType is standalone.
};
export const FieldType = {
  CALL_TYPE: 'CALL_TYPE',
  CHECKBOX: 'CHECKBOX',
  CHECKBOX_FIELD: 'CHECKBOX_FIELD',
  INTERMEDIATE: 'INTERMEDIATE',
  SELECT_SINGLE: 'SELECT_SINGLE',
  TAB: 'TAB',
  TEXT_BOX: 'TEXT_BOX',
  TEXT_INPUT: 'TEXT_INPUT',
};

export function isNotCategory(value) {
  const notCategory = ['error', 'touched', 'type', 'validation'];
  return notCategory.includes(value);
}

export function isNotSubcategory(value) {
  const notSubcategory = ['type'];
  return notSubcategory.includes(value);
}

// TODO: add tab order?

const defaultFormDefinition = {
  callType: {
    type: FieldType.CALL_TYPE,
  },
  callerInformation: {
    type: FieldType.TAB,
    name: {
      type: FieldType.INTERMEDIATE,
      firstName: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      lastName: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
    },
    relationshipToChild: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    gender: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    age: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    language: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    nationality: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    ethnicity: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    location: {
      type: FieldType.INTERMEDIATE,
      streetAddress: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      city: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      stateOrCounty: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      postalCode: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      phone1: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      phone2: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
    },
  },
  childInformation: {
    type: FieldType.TAB,
    name: {
      type: FieldType.INTERMEDIATE,
      firstName: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      lastName: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
    },
    gender: {
      type: FieldType.SELECT_SINGLE,
      validation: [ValidationType.REQUIRED],
    },
    age: {
      type: FieldType.SELECT_SINGLE,
      validation: [ValidationType.REQUIRED],
    },
    language: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    nationality: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    ethnicity: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    school: {
      type: FieldType.INTERMEDIATE,
      name: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      gradeLevel: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
    },
    location: {
      type: FieldType.INTERMEDIATE,
      streetAddress: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      city: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      stateOrCounty: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      postalCode: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      phone1: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
      phone2: {
        type: FieldType.TEXT_INPUT,
        validation: null,
      },
    },
    refugee: {
      type: FieldType.CHECKBOX,
      value: false,
    },
    disabledOrSpecialNeeds: {
      type: FieldType.CHECKBOX,
      value: false,
    },
    hiv: {
      type: FieldType.CHECKBOX,
      value: false,
    },
  },
  caseInformation: {
    type: FieldType.TAB,
    categories: {
      type: FieldType.CHECKBOX_FIELD,
      validation: [ValidationType.REQUIRED],
      'Missing children': {
        type: FieldType.INTERMEDIATE,
        'Child abduction': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Lost, injured or otherwise missing child': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        Runaway: {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      Violence: {
        type: FieldType.INTERMEDIATE,
        Bullying: {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Child / Early marriage': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Commercial sexual exploitation (offline)': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Economic exploitation': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Female Genital Mutilation (FGM)': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Gender-based harmful traditional practices (other than FGM)': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Harmful or hazardous labour': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Mental / Emotional violence': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Neglect (or negligent treatment)': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Non-Gender-Based harmful traditional practices': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Online sexual abuse': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Online sexual exploitation': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Physical violence': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Sexual violence': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      'Mental Health': {
        type: FieldType.INTERMEDIATE,
        'Addictive behaviours': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Concerns about the self': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Emotional distress – anger problems': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Emotional distress – fear and anxiety problems': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Emotional distress – mood problems': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Problems with eating behaviour': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Self-harming behaviours': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Substance use': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Suicidal thoughts and suicide attempts': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      'Physical Health': {
        type: FieldType.INTERMEDIATE,
        'Medical or lifestyle information about HIV/AIDS': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Medical problems': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Pregnancy and Maternal care': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Sexual and reproductive health': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      Accessibility: {
        type: FieldType.INTERMEDIATE,
        Education: {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Essential needs': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'General healthcare services': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Legal services and advice': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Mental health services': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Sexual health services': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Socio-economical services': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      'Discrimination and Exclusion': {
        type: FieldType.INTERMEDIATE,
        'Ethnicity/nationality': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Financial Situation': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Gender identity and expression/sexual identity': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        Health: {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Philosophical or religious beliefs': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      'Family Relationships': {
        type: FieldType.INTERMEDIATE,
        'Adoption, fostering, and extended family placement': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Relationship to parents': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Relationship with sibling(s)': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      'Peer Relationships': {
        type: FieldType.INTERMEDIATE,
        'Friends and Friendships': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Romantic Relationships': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      School: {
        type: FieldType.INTERMEDIATE,
        'Academic issues': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Teacher and school problems': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      Sexuality: {
        type: FieldType.INTERMEDIATE,
        'Sexuality and gender identity': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Sexual behaviours': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
      'Information & Other Non-Counselling contacts': {
        type: FieldType.INTERMEDIATE,
        'Complaints about the child helpline': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Questions by parents': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Questions about the child helpline': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Questions about other services': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        '"Thank you for your assistance"': {
          type: FieldType.CHECKBOX,
          value: false,
        },
        'Unspecified/Other': {
          type: FieldType.CHECKBOX,
          value: false,
        },
      },
    },
    callSummary: {
      type: FieldType.TEXT_BOX,
      validation: null,
    },
    referredTo: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    status: {
      value: 'In Progress',
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    keepConfidential: {
      type: FieldType.CHECKBOX,
      value: true,
    },
    okForCaseWorkerToCall: {
      type: FieldType.CHECKBOX,
      value: false,
    },
    howDidTheChildHearAboutUs: {
      type: FieldType.SELECT_SINGLE,
      validation: null,
    },
    didYouDiscussRightsWithTheChild: {
      type: FieldType.CHECKBOX,
      value: false,
    },
    didTheChildFeelWeSolvedTheirProblem: {
      type: FieldType.CHECKBOX,
      value: false,
    },
    wouldTheChildRecommendUsToAFriend: {
      type: FieldType.CHECKBOX,
      value: false,
    },
  },
  contactlessTask: {
    type: FieldType.INTERMEDIATE,
    channel: {
      type: FieldType.TEXT_INPUT,
      validation: null,
    },
    date: {
      type: FieldType.TEXT_INPUT,
      validation: null,
    },
    time: {
      type: FieldType.TEXT_INPUT,
      validation: null,
    },
  },
};

const recursivelyCreateBlankForm = formDefinition => {
  const initialState = {};

  Object.keys(formDefinition)
    .filter(key => key !== 'type' && key !== 'validation')
    .forEach(key => {
      switch (formDefinition[key].type) {
        case FieldType.CALL_TYPE:
          initialState[key] = {
            ...formDefinition[key],
            value: '',
          };
          break;
        case FieldType.CHECKBOX:
          initialState[key] = {
            value: false, // set default of false if not overridden
            ...formDefinition[key],
          };
          break;
        case FieldType.CHECKBOX_FIELD:
          initialState[key] = {
            ...recursivelyCreateBlankForm(formDefinition[key]),
            type: formDefinition[key].type,
            touched: false,
            validation: formDefinition[key].validation === null ? null : Array.from(formDefinition[key].validation),
            error: null,
          };
          break;
        case FieldType.INTERMEDIATE:
          initialState[key] = {
            ...recursivelyCreateBlankForm(formDefinition[key]),
            type: formDefinition[key].type,
          };
          break;
        case FieldType.TAB:
          initialState[key] = {
            ...recursivelyCreateBlankForm(formDefinition[key]),
            type: formDefinition[key].type,
          };
          break;
        case FieldType.SELECT_SINGLE:
        case FieldType.TEXT_BOX:
        case FieldType.TEXT_INPUT:
          initialState[key] = {
            value: '', // set default of empty if not overridden
            ...formDefinition[key],
            touched: false,
            error: null,
          };
          break;
        default:
          throw new Error(
            `Unknown FieldType ${formDefinition[key].type} for key ${key} in ${JSON.stringify(formDefinition)}`,
          );
      }
    });

  return initialState;
};

const createCategoriesMetadata = formDef => {
  const expanded = Object.keys(formDef.caseInformation.categories)
    .filter(key => !isNotCategory(key))
    .reduce((result, key) => ({ ...result, [key]: false }), {});

  return {
    gridView: false,
    expanded,
  };
};

/**
 * @param {any} formDef
 * The form definition we want to replicate.
 * @param {Boolean} recreated
 * Determines if we are initializing a form or recreating it
 * @returns {any}
 * A clean form plus metadata about it.
 * If the form was recreated, it won't have a "starting time".
 */
export const createBlankForm = (formDef = defaultFormDefinition, recreated = false) => {
  const initialState = recursivelyCreateBlankForm(formDef);
  const metadata = {
    startMillis: recreated ? null : new Date().getTime(),
    endMillis: null,
    tab: 1,
    recreated,
    categories: createCategoriesMetadata(formDef),
  };

  const generatedForm = { ...initialState, metadata };

  return generatedForm;
};

export const recreateBlankForm = () => {
  const recreatedForm = createBlankForm(defaultFormDefinition, true);

  return recreatedForm;
};
