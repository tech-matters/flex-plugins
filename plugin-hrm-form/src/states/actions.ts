import {
  GeneralActionType,
  INITIALIZE_CONTACT_STATE,
  RECREATE_CONTACT_STATE,
  REMOVE_CONTACT_STATE,
  ContactFormDefinition,
} from './types';

export const initializeContactState = (definitions: ContactFormDefinition) => (taskId: string): GeneralActionType => ({
  type: INITIALIZE_CONTACT_STATE,
  definitions,
  taskId,
});

export const recreateContactState = (definitions: ContactFormDefinition) => (taskId: string): GeneralActionType => ({
  type: RECREATE_CONTACT_STATE,
  definitions,
  taskId,
});

export const removeContactState = (taskId: string): GeneralActionType => ({
  type: REMOVE_CONTACT_STATE,
  taskId,
});
