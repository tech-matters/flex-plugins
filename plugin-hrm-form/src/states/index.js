import { combineReducers } from 'redux';

import { reduce as ContactStateReducer } from './ContactState';
import { reduce as SearchFormReducer } from './SearchContact';
import { reduce as ConnectedCaseReducer } from './case/reducer';
import { reduce as QueuesStatusReducer } from './queuesStatus/reducer';
import { reduce as ConfigurationReducer } from './configuration/reducer';
import { reduce as RoutingReducer } from './routing/reducer';

// Register your redux store under a unique namespace
export const namespace = 'plugin-hrm-form';
export const contactFormsBase = 'activeContacts';
export const searchContactsBase = 'searchContacts';
export const connectedCaseBase = 'connectedCase';
export const queuesStatusBase = 'queuesStatusState';
export const configurationBase = 'configuration';
export const routingBase = 'routing';

// Combine the reducers
export default combineReducers({
  [contactFormsBase]: ContactStateReducer,
  [searchContactsBase]: SearchFormReducer,
  [queuesStatusBase]: QueuesStatusReducer,
  [connectedCaseBase]: ConnectedCaseReducer,
  [configurationBase]: ConfigurationReducer,
  [routingBase]: RoutingReducer,
});
