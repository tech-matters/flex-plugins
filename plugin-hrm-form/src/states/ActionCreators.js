import { HANDLE_BLUR, HANDLE_FOCUS, SAVE_CONTACT_STATE } from './ActionTypes';
import { formIsValid, moreThanThreeCategoriesSelected, validateBeforeSubmit, validateOnBlur } from './ValidationRules';

export const handleBlur = dispatch => (form, taskId) => () => {
  console.log('Received blur event');
  dispatch({
    type: HANDLE_BLUR,
    form: validateOnBlur(form),
    taskId,
  });
};

export const handleFocus = dispatch => (taskId, parents, name) => {
  console.log(`handleFocus called with taskId = ${taskId}, parents = ${JSON.stringify(parents)}, name = ${name}`);
  dispatch({
    type: HANDLE_FOCUS,
    parents,
    name,
    taskId,
  });
};

export const handleSubmit = dispatch => (form, hrmBaseUrl, workerSid, helpline, handleCompleteTask) => task => {
  const newForm = validateBeforeSubmit(form);
  dispatch({
    type: HANDLE_BLUR, // probably need to rename this
    form: newForm,
    taskId: task.taskSid,
  });
  if (formIsValid(newForm)) {
    dispatch({
      type: SAVE_CONTACT_STATE,
      hrmBaseUrl,
      task,
      abortFunction: () => null,
      workerSid,
      helpline,
    });
    handleCompleteTask(task.taskSid, task);
  } else {
    window.alert('There is a problem with your submission.  Please check the form for errors.');
  }
};

/*
 * This is not technically an ActionCreator.  It's more of a filter and/or validator,
 * but it happens on an event.  Does it belong here?
 * This is also (dangerously?) hardcoding where the category information is.
 */
export const handleCategoryToggle = (form, handleChange) => (taskId, category, subcategory, newValue) => {
  const candidateCategories = {
    ...form.caseInformation.categories,
    [category]: {
      ...form.caseInformation.categories[category],
      [subcategory]: {
        ...form.caseInformation.categories[category][subcategory],
        value: newValue,
      },
    },
  };
  if (moreThanThreeCategoriesSelected(candidateCategories)) {
    window.alert('You cannot select more than three category options');
  } else {
    handleChange(taskId, ['caseInformation', 'categories', category], subcategory, newValue);
  }
};
