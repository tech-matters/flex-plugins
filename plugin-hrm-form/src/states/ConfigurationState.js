import { CHANGE_LANGUAGE } from './ActionTypes';

// default language to initialize plugin
export const defaultLanguage = 'en-US';

export const changeLanguage = language => ({ type: CHANGE_LANGUAGE, language });

const initialState = {
  language: defaultLanguage,
};

export function reduce(state = initialState, action) {
  switch (action.type) {
    case CHANGE_LANGUAGE:
      return {
        ...state,
        language: action.language,
      };
    default:
      return state;
  }
}
