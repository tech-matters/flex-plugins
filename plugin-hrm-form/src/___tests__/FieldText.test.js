import React from 'react';
import renderer from 'react-test-renderer';

import FieldText from '../components/FieldText';
import { ValidationType } from '../states/ContactFormStateFactory';

jest.mock('../Styles/HrmStyles', () => {
  return {
    StyledInput: 'StyledInput',
    StyledLabel: 'StyledLabel',
    ErrorText: 'ErrorText',
    TextField: 'TextField',
  };
});

// TODO: Improve assertion for all tests below
test('render basic FieldText', () => {
  const firstName = {
    value: '',
    touched: false,
    error: null,
    validation: null,
  };
  const handleBlur = jest.fn();
  const handleChange = jest.fn();
  const handleFocus = jest.fn();
  const taskId = '';

  const component = renderer.create(
    <FieldText
      id="CallerInformation_FirstName"
      label="First name"
      parents={['callerInformation', 'name']}
      name="firstName"
      field={firstName}
      taskId={taskId}
      handleBlur={handleBlur}
      handleChange={handleChange}
      handleFocus={handleFocus}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render required FieldText', () => {
  const firstName = {
    value: '',
    touched: false,
    error: null,
    validation: [ValidationType.REQUIRED],
  };
  const handleBlur = jest.fn();
  const handleChange = jest.fn();
  const handleFocus = jest.fn();
  const taskId = '';

  const component = renderer.create(
    <FieldText
      id="CallerInformation_FirstName"
      label="First name"
      parents={['callerInformation', 'name']}
      name="firstName"
      field={firstName}
      taskId={taskId}
      handleBlur={handleBlur}
      handleChange={handleChange}
      handleFocus={handleFocus}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render FieldText with errors', () => {
  const firstName = {
    value: '',
    touched: true,
    error: 'This field is required',
    validation: [ValidationType.REQUIRED],
  };
  const handleBlur = jest.fn();
  const handleChange = jest.fn();
  const handleFocus = jest.fn();
  const taskId = '';

  const component = renderer.create(
    <FieldText
      id="CallerInformation_FirstName"
      label="First name"
      parents={['callerInformation', 'name']}
      name="firstName"
      field={firstName}
      taskId={taskId}
      handleBlur={handleBlur}
      handleChange={handleChange}
      handleFocus={handleFocus}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
