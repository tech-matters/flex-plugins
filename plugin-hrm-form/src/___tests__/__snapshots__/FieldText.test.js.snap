// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render FieldText with errors 1`] = `
<TextField
  name="firstName"
  parents={
    Array [
      "callerInformation",
      "name",
    ]
  }
  taskId=""
>
  <StyledLabel
    htmlFor="CallerInformation_FirstName"
  >
    First name
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      *
    </span>
  </StyledLabel>
  <StyledInput
    error={true}
    id="CallerInformation_FirstName"
    multiline={false}
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder=""
    value=""
  />
  <ErrorText>
    This field is required
  </ErrorText>
</TextField>
`;

exports[`render basic FieldText 1`] = `
<TextField
  name="firstName"
  parents={
    Array [
      "callerInformation",
      "name",
    ]
  }
  taskId=""
>
  <StyledLabel
    htmlFor="CallerInformation_FirstName"
  >
    First name
  </StyledLabel>
  <StyledInput
    error={false}
    id="CallerInformation_FirstName"
    multiline={false}
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder=""
    value=""
  />
</TextField>
`;

exports[`render required FieldText 1`] = `
<TextField
  name="firstName"
  parents={
    Array [
      "callerInformation",
      "name",
    ]
  }
  taskId=""
>
  <StyledLabel
    htmlFor="CallerInformation_FirstName"
  >
    First name
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      *
    </span>
  </StyledLabel>
  <StyledInput
    error={false}
    id="CallerInformation_FirstName"
    multiline={false}
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder=""
    value=""
  />
</TextField>
`;
